### PR TITLE
Client Charge pagination changes

### PR DESCRIPTION
**Pull request by VenuShyam**

### CODE REVIEW

File: mifosng-provider/src/integrationTest/java/org/mifosplatform/integrationtests/common/ClientChargesTest.java [hunk 1]
```diff
@@ -0,0 +1,121 @@
+package org.mifosplatform.integrationtests.common;
```
Comment (vishwasbabu) [inline mifosng-provider/src/integrationTest/java/org/mifosplatform/integrationtests/common/ClientChargesTest.java]: Licences need to be applied , refer https://github.com/openMF/mifosx/blob/develop/Contributing.md point 10

File: mifosng-provider/src/integrationTest/java/org/mifosplatform/integrationtests/common/ClientChargesTest.java [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+    @Before
+    public void setup() {
+        Utils.initializeRESTAssured();
+        this.requestSpec = new RequestSpecBuilder().setContentType(ContentType.JSON).build();
+        this.requestSpec.header("Authorization", "Basic " + Utils.loginIntoServerAndGetBase64EncodedAuthenticationKey());
+        this.responseSpec = new ResponseSpecBuilder().expectStatusCode(200).build();
+    }
+    @Test
+    public void createClientCharge_Success(){
```
Comment (vishwasbabu) [inline mifosng-provider/src/integrationTest/java/org/mifosplatform/integrationtests/common/ClientChargesTest.java]: do not use _ in method names

File: mifosng-provider/src/integrationTest/java/org/mifosplatform/integrationtests/common/ClientChargesTest.java [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+       
+    }
+    //undo transaction
+   private void isUndoTransaction(String clientId,String transactionId){
+       final Boolean isReversed=ClientHelper.getClientTransactions(this.requestSpec, this.responseSpec,clientId.toString(), 
+               transactionId);
+       Assert.assertTrue(isReversed);
+   }
+   
+   private void isValidOutstandingAmount(Object outStandingAmount,Object expectedAmount){
```
Comment (vishwasbabu) [inline mifosng-provider/src/integrationTest/java/org/mifosplatform/integrationtests/common/ClientChargesTest.java]: Files need to be formatted using our formatter , refer https://github.com/openMF/mifosx/blob/develop/Contributing.md point 6

File: mifosng-provider/src/main/java/org/mifosplatform/portfolio/calendar/service/CalendarUtils.java [hunk 4]
```diff
@@ -111,13 +111,11 @@ private static Date convertToiCal4JCompatibleDate(final LocalDate inputDate) {
     private static Collection<LocalDate> convertToLocalDateList(final DateList dates) {
 
         final Collection<LocalDate> recurringDates = new ArrayList<>();
```
Comment (vishwasbabu) [inline mifosng-provider/src/main/java/org/mifosplatform/portfolio/calendar/service/CalendarUtils.java]: please make the calendar changes a seperate pull request. Best to have one pr correspond to one issue on JIra

File: mifosng-provider/src/main/java/org/mifosplatform/portfolio/client/service/ClientReadPlatformServiceImpl.java [hunk 5]
```diff
@@ -210,6 +210,7 @@ public ClientData retrieveOne(final Long clientId) {
 
             final String sql = "select " + this.clientMapper.schema()
                     + " where ( o.hierarchy like ? or transferToOffice.hierarchy like ?) and c.id = ?";
+
```
Comment (vishwasbabu) [inline mifosng-provider/src/main/java/org/mifosplatform/portfolio/client/service/ClientReadPlatformServiceImpl.java]: unchanged files should not be a part of pr

File: mifosng-provider/src/main/java/org/mifosplatform/portfolio/client/service/ClientTransactionReadPlatformService.java [hunk 6]
```diff
@@ -5,18 +5,18 @@
...
 
 public interface ClientTransactionReadPlatformService {
 
     @Transactional(readOnly = true)
-    public Collection<ClientTransactionData> retrieveAllTransactions(Long clientId);
+    public Page<ClientTransactionData> retrieveAllTransactions(Long clientId,Integer limit,Integer offset);
 
     @Transactional(readOnly = true)
-    public Collection<ClientTransactionData> retrieveAllTransactions(final Long clientId, final Long chargeId);
+    public Page<ClientTransactionData> retrieveAllTransactions(final Long clientId, final Long chargeId,Integer limit,Integer offset);
```
Comment (vishwasbabu) [inline mifosng-provider/src/main/java/org/mifosplatform/portfolio/client/service/ClientTransactionReadPlatformService.java]: no need to paginate all transactions paying a particular charge

File: mifosng-provider/src/main/java/org/mifosplatform/portfolio/loanaccount/loanschedule/domain/DefaultScheduledDateGenerator.java [hunk 7]
```diff
@@ -60,7 +59,8 @@ public LocalDate generateNextRepaymentDate(final LocalDate lastRepaymentDate, fi
                 // repayment
                 LocalDate seedDate = currentCalendar.getStartDateLocalDate();
                 String reccuringString = currentCalendar.getRecurrence();
-                dueRepaymentPeriodDate = CalendarUtils.getNewRepaymentMeetingDate(reccuringString, seedDate, dueRepaymentPeriodDate,
+                // dueRepaymentPeriodDate = null; which is causing the exception in isValidRecurringdates() and resulting in not updating the repayment schedule
+                dueRepaymentPeriodDate = CalendarUtils.getNewRepaymentMeetingDate(reccuringString, seedDate, lastRepaymentDate,
```
Comment (vishwasbabu) [inline mifosng-provider/src/main/java/org/mifosplatform/portfolio/loanaccount/loanschedule/domain/DefaultScheduledDateGenerator.java]: please make this a separate pull request

**Comment by vishwasbabu** [comment]
Can you please address the review comments and resend the pr along with updated API documentation

**Comment by VenuShyam** [comment]
Hi Vishwas ,

    I have done the changes as per the review comments...I have tried a lot
to send two separate pull requests but I failed to achieve it,almost I have
spent 3 hours to delete the two separate commits and send them as one.As I
have done the two commits from two different branches I could not able to
resolve it and produce as single commit and separate PR .I have checked
with the fellow mates also but no body had an idea about it.Kindly excuse
me for this time and once you are in the office I will get the knowledge
from you and I will ensure that it will not repeat in the future.




Thanks,
Venu,
Software Engineer,
Conflux Technologies,
Bangalore-43
